modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+                        // Copy actual row id into a "seekKeyBuf" buffer.
+                        GridUnsafe.copyMemory(
+                                null, GridUnsafe.bufferAddress(directBuffer) + ROW_ID_OFFSET,
+                                seekKeyBuf.array(), GridUnsafe.BYTE_ARR_OFF + ROW_ID_OFFSET,
+                                igniteRowIdSize
+                        );
+
+                        // Perform additional "seek" if timestamp is not null. Motivation for it is described in comments above.
+                        if (timestamp != null) {
+                            // At this point, "seekKeyBuf" has row id that exists in partition.
+                            it.seek(seekKeyBuf.array());
+
+                            // Iterator may not be valid if that row was created after required timestamp.
+                            if (invalid(it)) {
+                                return false;
+                            }
+
+                            // Or iterator may still be valid even if there's no version for required timestamp. In this case row id
+                            // itself will be different and must check it.
+                            it.key(directBuffer);
+
+                            if (!directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
+                                found = false;
+                            }
+                        }
+                    }
+
+                    // This one might look tricky. We finished processing next row. There are three options:
+                    //  - "found" flag is false - there's no fitting version of the row. We'll continue to next iteration;
+                    //  - value is empty, we found a tombstone. We'll continue to next iteration as well;
+                    //  - value is not empty and everything's good. We'll cache it and return from method.
+                    // In all three cases we need to prepare the value of "seekKeyBuf" so that it has not-yet-scanned row id in it.
+                    // the only valid way to do so is to treat row id payload as one big unsigned integer in Big Endian and increment it.
+                    // It's important to note that increment may overflow. In this case "carry flag" will go into incrementing partition id.
+                    // This is fine for three reasons:
+                    //  - iterator has an upper bound, following "seek" will result in invalid iterator state.
+                    //  - partition id iself cannot be overflown, because it's limited with a constant less than 0xFFFF. It's something like
+                    //    65500, I think.
+                    //  - "seekKeyBuf" buffer value will not be used after that, so it's ok if we corrupt its data (in every other instance,
+                    //    buffer starts with a valid partition id, which is set during buffer's initialization).
+                    for (int i = ROW_ID_OFFSET + igniteRowIdSize - 1;; i--) {
```

Comment:
Let's extract this incrementation to a method of its own

Reply:
Ok